### PR TITLE
Use file paths/urls correctly

### PR DESCRIPTION
**Pull request by haampie**
The main issue that's fixed is that Spack passes paths (as strings) to
functions that require urls. That wasn't an issue on unix, since there
you can simply concatenate `file://` and `path` and all is good, but on
Windows that gives invalid file urls.

Instead of all sorts of ad-hoc `if windows: fix_broken_file_url` this PR
adds two helper functions around Python's own path2url and reverse.
`spack.util.url.path_to_file_url` / `spack.util.url.file_url_string_to_path`

This allows me to remove the custom `spack.util.url.parse` function.
That function is generally broken, as it has its own interpretation of
the `file://` protocol: it tries to be smart about relative paths, but the
truth is that this is simply not supported by design, so any attempt
to make `file://x/y` mean the relative local path `x/y` is plain wrong.
This URI has a meaning: `hostname=x` and `path=/y`.

Also the variable interpolation in `spack.util.url.parse` has similar
issues, where Spack just tries to be too smart, disagreeing with
the specification, and not particularly following the principle of
least surprise. Just let the call-site interpolate variables if necessary.

Also fixes what I think is a bug, where some `spack buildcache` commands
used `-d` as a flag to mean `--mirror-url` requiring a URL, and others
`--directory`, requiring a path. (For @scottwittenburg)

Adding @tldahlgren as a reviewer for the environment include URL
stuff; fyi it also fixes a bug where a non-existing local path would
result in a FetchError, which is weird. See the updated test.

### CODE REVIEW

**Comment by haampie** [comment]
@spackbot fix style

**Comment by haampie** [comment]
@johnwparent can we get rid of the "custom" urlparse? To me it's the wrong approach.

File: lib/spack/spack/util/url.py
Comment (johnwparent) [inline lib/spack/spack/util/url.py:374]: Are we still enforcing this transitively through `urllib`? I.e. `urllib` will throw if given anything but an expected url and the error message is clear, or does this refactor mean this is no longer necessary and we're just converting any potential non url strings into urls? I think it's the latter but it's not immediately obvious.

File: lib/spack/spack/test/cmd/env.py
Comment (johnwparent) [inline lib/spack/spack/test/cmd/env.py:712]: This test isn't currently run on Windows but thanks for fixing it anyway!

**Comment by johnwparent** [comment]
Overall this looks fine. I left one comment, which is fairly minor and does not need to hold up the progress of this branch. Testing locally it looks like the `urllib` utilities provides at least the same functionality (likely better) than our custom url parsing and we should be moving away from our own url parsing anyway, so I think it's perfectly reasonable to remove the custom urlparse.

File: lib/spack/spack/util/url.py
Comment (haampie) [reply lib/spack/spack/util/url.py:374]: urllib won't throw during parsing. My thinking was that the parser can't be a validator, cause what scheme is supported depends on the call site. For buildcache etc we still have some validation that at least errors when a url is passed to `--directory` and vice versa, although I find it a bit pedantic to be honest ;p 

It's easier to validate after parsing, e.g. `assert url.scheme in supported_schemes`. Are there any obvious places where that should be added? For fetching at some point you'll just get a URLError such as `URLError: <urlopen error unknown url type: unknown>`.

File: lib/spack/spack/cmd/buildcache.py
Comment (tldahlgren) [inline lib/spack/spack/cmd/buildcache.py]: I know these are mutually exclusive, but why not `elif` here and at line 722?

Comment (haampie) [reply lib/spack/spack/cmd/buildcache.py]: It was copied from elsewhere, but sounds good to me

**Comment by haampie** [comment]
@johnwparent can we merge this rather sooner than later? Then I can move on with #34498

Also @johnwparent, how about hostnames in file urls? E.g. `file://hostname/x` has `netloc=hostname` and `path=/x`. I don't think it should be part of this PR, since Spack erroneously interpreted that file url as a relative path `hostname/x` in before, so it was incorrect anyways.

**Comment by johnwparent** [comment]
> @johnwparent can we merge this rather sooner than later? Then I can move on with #34498

I think this is ok to merge now.

> Also @johnwparent, how about hostnames in file urls? E.g. `file://hostname/x` has `netloc=hostname` and `path=/x`. I don't think it should be part of this PR, since Spack erroneously interpreted that file url as a relative path `hostname/x` in before, so it was incorrect anyways.

I'm inclined to agree with you here. I've started a secondary effort to retool our approach to urls/paths for better cross platform support and I can handle this there. This is blocking your other works so LGTM.

**Comment by haampie** [comment]
Actually it's triggering a problem in CI where (probably) a `urlparse(...)` object is fed to `urlparse(...)`. Apparently that's not tested by unit tests

**Comment by johnwparent** [comment]
> Actually it's triggering a problem in CI where (probably) a `urlparse(...)` object is fed to `urlparse(...)`. Apparently that's not tested by unit tests

Huh, I didn't notice a point where that could happen, surprising we don't test for something like that. 

If you don't want to add that test here, and just fix the issue, I'll add a to do to my effort to update that test suite.

**Comment by haampie** [comment]
Apparently CI did not run on the right commit :(

1f627df did [run here](https://gitlab.spack.io/spack/spack/-/pipelines/238355), which actually ran on this [merge commit](https://gitlab.spack.io/spack/spack/-/commit/60e7ed08dc11ecba297f1b93acc3c3d67c2a54e9), which has e0ab93226264d3218545daefc876da17142ee413 as a parent instead of 1f627df :|

I tried to rebase on 94bac8d6dd9140820dbde9d1dda4ea6c5736ca43 to see if it syncs better. Ping @zackgalbreath / @scottwittenburg

**Comment by scottwittenburg** [comment]
> which has [e0ab932](https://github.com/spack/spack/commit/e0ab93226264d3218545daefc876da17142ee413) as a parent instead of [1f627df](https://github.com/spack/spack/commit/1f627dfe40880231848588c2eab6c3efddeed320)

Hrm, yeah, there's definitely something wrong there.  The commit message has the right parents (including the PR HEAD sha we wanted to test), but it doesn't match the actual parents of the commit.

**Comment by johnwparent** [comment]
> > which has [e0ab932](https://github.com/spack/spack/commit/e0ab93226264d3218545daefc876da17142ee413) as a parent instead of [1f627df](https://github.com/spack/spack/commit/1f627dfe40880231848588c2eab6c3efddeed320)
> 
> Hrm, yeah, there's definitely something wrong there. The commit message has the right parents (including the PR HEAD sha we wanted to test), but it doesn't match the actual parents of the commit.

I think this may somehow be a GH issue, I've seen RTD pulling the wrong commit a few times as well, and have heard complaints from other users as well today.

**Comment by scottwittenburg** [comment]
So the sync script does have two sources of information that need to jive on the GH end.  The branch we create starts, basically, with:

```
git fetch github refs/pull/<pull_number>/head:<tmp_branch>
```

Then later we merge `<tmp_branch>` into a new branch based on the latest tested `develop`.  So that's the source of the *actual* parents of the merge commit we make.  On the other hand, we get one of the parents of the commit *message*  we push from a GH api endpoint that provides the pull request head sha.

If those two sources of information don't agree, then we're in trouble.  But I'm looking to see if there's a way we can be immune to a mismatch there, assuming that's what happened.

**Comment by haampie** [comment]
> I think this may somehow be a GH issue, I've seen RTD pulling the wrong commit a few times as well, and have heard complaints from other users as well today.

Yeah, it's likely a Github issue.

https://readthedocs.org/projects/spack/builds/18906514/

Notice: **Failed to checkout revision: 1f627dfe40880231848588c2eab6c3efddeed320**

But then it should also really be a Gitlab CI failure instead of running on the wrong/old commit.

**Comment by haampie** [comment]
Should we run a spackbot-rebuild-everything on this?

**Comment by haampie** [comment]
@spackbot rebuild everything

**Comment by haampie** [comment]
Seems like everything is on fire today :(

File: lib/spack/spack/environment/environment.py
Comment (tldahlgren) [inline lib/spack/spack/environment/environment.py:936]: What is an example of a case where `config_path` != `include_url.path` here?

**Comment by haampie** [comment]
@spackbot run pipeline

File: lib/spack/spack/environment/environment.py
Comment (tldahlgren) [inline lib/spack/spack/environment/environment.py:946]: Same question here about `remote_path` != `include_url.path`.

Comment (tldahlgren) [reply lib/spack/spack/environment/environment.py:946]: Also, IMO the variable name is misleading.  Is this not a local path?

**Review by tldahlgren — COMMENTED** [review]
At this point looked at the environment change as requested offline.  Have some questions.

File: lib/spack/spack/environment/environment.py
Comment (tldahlgren) [reply lib/spack/spack/environment/environment.py:936]: Is this needed for `windows`?

File: lib/spack/spack/test/cmd/env.py
Comment (tldahlgren) [reply lib/spack/spack/test/cmd/env.py:712]: Great catch btw!

File: lib/spack/spack/environment/environment.py
Comment (tldahlgren) [reply lib/spack/spack/environment/environment.py:946]: BUT, I can adjust when I resolve conflicts with one of my other environment-related PRs.

Comment (tldahlgren) [reply lib/spack/spack/environment/environment.py:946]: BUT I can adjust the name after addressing merge conflicts in one of my other environment-related PRs.

**Comment by haampie** [comment]
So, gitlab rebuild a few things during rebuild everything, but the vast majority failed with unrelated issues. I think it's good enough.

Now I'm praying that it won't take 10 attempts to make gitlab ci pass on nothing to rebuild.

@spackbot run pipeline

File: lib/spack/spack/environment/environment.py
Comment (haampie) [reply lib/spack/spack/environment/environment.py:936]: Windows because `file:///C:/x` would give a path `/C:/x` which url2pathname transforms into `C:\\x`.

But also on Linux it decodes the url, e.g. `file:///hello%20world` -> `/hello world`.

File: lib/spack/spack/environment/environment.py
Comment (haampie) [reply lib/spack/spack/environment/environment.py:946]: Yeah, that sounds like a better name

**Comment by haampie** [comment]
@spackbot run pipeline

:|

**Comment by haampie** [comment]
@spackbot run pipeline

**Comment by haampie** [comment]
@spackbot run pipeline

**Comment by haampie** [comment]
@spackbot run pipeline

**Comment by haampie** [comment]
Finally it's in a merge-able state again! Pinging @johnwparent one more time

**Review by johnwparent — APPROVED** [review]
Looks good from the Windows side of things!

**Pull request merged**
